fineract-client/src/test/java/org/apache/fineract/integrationtests/newstyle/DocumentTest.java

```diff
@@ -0,0 +1,114 @@
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.fineract.integrationtests.newstyle;
+
+import java.io.File;
+import java.io.IOException;
+import okhttp3.MediaType;
+import okhttp3.MultipartBody.Part;
+import okhttp3.ResponseBody;
+import org.apache.fineract.client.models.GetEntityTypeEntityIdDocumentsResponse;
+import org.apache.fineract.client.models.PostEntityTypeEntityIdDocumentsResponse;
+import org.apache.fineract.client.util.Parts;
+import org.junit.jupiter.api.Order;
+import org.junit.jupiter.api.Test;
+
+/**
+ * Integration Test for /documents API.
+ *
+ * @author Michael Vorburger.ch
+ */
+public class DocumentTest extends IntegrationTest {
+
+    final File testFile = new File(getClass().getResource("/michael.vorburger-crepes.jpg").getFile());
+
+    Long clientId = new ClientTest().getClientId();
+    Long documentId;
+
+    @Test
+    @Order(1)
+    void retrieveAllDocuments() throws IOException {
+        assertThat(ok(fineract().documents.retreiveAllDocuments("clients", clientId))).isNotNull();
+    }
+
+    @Test
+    @Order(2)
+    void createDocument() throws IOException {
+        String name = "Test";
+        Part part = Parts.fromFile(testFile);
+        String description = null;
+        // TODO used var in tests when moved from fineract-client (Java 8 only) to new module
+        PostEntityTypeEntityIdDocumentsResponse response = ok(
+                fineract().documents.createDocument("clients", clientId, part, name, description));
+        assertThat(response.getResourceId()).isNotNull();
+        assertThat(response.getResourceIdentifier()).isNotEmpty();
+        documentId = response.getResourceId();
+    }
+
+    @Test
+    @Order(3)
+    void getDocument() throws IOException {
+        GetEntityTypeEntityIdDocumentsResponse doc = ok(fineract().documents.getDocument("clients", clientId, documentId));
+        assertThat(doc.getName()).isEqualTo("Test");
+        assertThat(doc.getFileName()).isEqualTo(testFile.getName());
+        assertThat(doc.getDescription()).isNull();
+        assertThat(doc.getId()).isEqualTo(documentId);
+        assertThat(doc.getParentEntityType()).isEqualTo("clients");
+        assertThat(doc.getParentEntityId()).isEqualTo(clientId);
+        // TODO huh?! It's more than uploaded file; seems like a bug - it's including create body, not just file size
+        assertThat(doc.getSize()).isEqualTo(testFile.length() + 385);
+        // TODO huh?! MIME is always text/plain instead of image/jpeg... :(
+        assertThat(doc.getType()).isEqualTo("text/plain");
+        // TODO doc.getStorageType() shouldn't be exposed by the API?!
+    }
```

Comment:
Brilliant to see how many bugs writing a decent test uncovers! :-)

Reply:
Indeed. As well as how easy it now is to write decent tests, using the shiny new client API! PS: The MIME thing is a bug in the test util, not core - fix for that coming up as soon as this PR is merged... :smile_cat: